### ISSUE TITLE
Implement RFC-007 for all providers

### ISSUE THREAD

**Issue by dipti-pai** [fluxcd/flux2#4846]
### Describe the bug

This issue tracks the implementation of [RFC-007](https://github.com/fluxcd/flux2/tree/main/rfcs/0007-git-repo-passwordless-auth)

### Steps to reproduce

# WorkItems

## Implement pkg changes

- [ ] Implement azure auth pkg to obtain workload identity credentials to access ADO
- [ ] Implement aws auth pkg to obtain sts credentials to access CodeCommit
- [ ] Implement gcp auth pkg to fetch the access token to access CSR
- [ ] Implement github provider to read the provided secret and get the gh app installation token
- [ ] Implement gitlab provider to read the provided secret and get the access token

## Controller changes

- [ ] Add .spec.provider to GitRepository API
- [ ] Implement source-controller changes to use the provider to authenticate to git
- [ ] Implement image-automation-controller changes to use the provider to authenticate to git

### Expected behavior

-

### OS / Distro

-

### Flux version

-

### Flux check

-

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by stefanprodan** [comment]
I have removed `CodeCommit` as this service has been shutdown by AWS.

**Comment by qzmiro** [comment]
Hello! 
The issue states for `all controllers` but does this include the `notification-controller`? i'm trying to use the **GithubDispatch** provider in my company but since it allows only PAT token, it's a bit complicated to negociate with the security team.

**Comment by stefanprodan** [comment]
@qzmiro I agree that notification-controller should work with GitHub App auth for status updates and workflow dispatch. Can you please open an issue in NC repo for these two?

**Comment by matheuscscp** [comment]
@dipti-pai Is there any items left from this issue? I marked caching as done, and I suppose GCP is no longer in the scope of this issue, so maybe only GitLab left?
